tests/unit-tests/test-util.cc

```diff
@@ -0,0 +1,594 @@
+    }
+
+
+    TEST(absPath, pathIsCanonicalised) {
+        auto path = "/some/path/with/trailing/dot/.";
+        auto p1 = absPath(path);
+        auto p2 = absPath(p1);
+
+        ASSERT_STREQ(p1.c_str(), "/some/path/with/trailing/dot");
+        ASSERT_STREQ(p1.c_str(), p2.c_str());
+    }
+
+    /* ----------------------------------------------------------------------------
+     * canonPath
+     * --------------------------------------------------------------------------*/
+
+    TEST(canonPath, removesTrailingSlashes) {
+        auto path = "/this/is/a/path//";
+        auto p = canonPath(path);
+
+        ASSERT_STREQ(p.c_str(), "/this/is/a/path");
+    }
+
+    TEST(canonPath, removesDots) {
+        auto path = "/this/./is/a/path/./";
+        auto p = canonPath(path);
+
+        ASSERT_STREQ(p.c_str(), "/this/is/a/path");
+    }
+
+    TEST(canonPath, removesDots2) {
+        auto path = "/this/a/../is/a////path/foo/..";
+        auto p = canonPath(path);
+
+        ASSERT_STREQ(p.c_str(), "/this/is/a/path");
+    }
+
+    TEST(canonPath, requiresAbsolutePath) {
+        ASSERT_ANY_THROW(canonPath("."));
+        ASSERT_ANY_THROW(canonPath(".."));
+        ASSERT_ANY_THROW(canonPath("../"));
+        ASSERT_DEATH({ canonPath(""); }, "path != \"\"");
+    }
+
+    /* ----------------------------------------------------------------------------
+     * dirOf
+     * --------------------------------------------------------------------------*/
+
+    // XXX: according to the doc of `dirOf`, dirOf("/") and dirOf("/foo")
+    // should both return "" but it actually returns "/" in both cases
+    TEST(dirOf, DISABLED_returnsEmptyStringForRoot) {
+        auto p = dirOf("/");
+
+        ASSERT_STREQ(p.c_str(), "");
+    }
+
+    TEST(dirOf, returnsFirstPathComponent) {
+        auto p1 = dirOf("/dir/");
+        ASSERT_STREQ(p1.c_str(), "/dir");
+        auto p2 = dirOf("/dir");
+        ASSERT_STREQ(p2.c_str(), "/");
+        auto p3 = dirOf("/dir/..");
+        ASSERT_STREQ(p3.c_str(), "/dir");
+        auto p4 = dirOf("/dir/../");
+        ASSERT_STREQ(p4.c_str(), "/dir/..");
+    }
+
+    /* ----------------------------------------------------------------------------
+     * baseNameOf
+     * --------------------------------------------------------------------------*/
+
+    TEST(baseNameOf, emptyPath) {
+        auto p1 = baseNameOf("");
+        ASSERT_STREQ(std::string(p1).c_str(), "");
+    }
+
+    TEST(baseNameOf, pathOnRoot) {
+        auto p1 = baseNameOf("/dir");
+        ASSERT_STREQ(std::string(p1).c_str(), "dir");
+    }
+
+    TEST(baseNameOf, relativePath) {
+        auto p1 = baseNameOf("dir/foo");
+        ASSERT_STREQ(std::string(p1).c_str(), "foo");
+    }
+
+    TEST(baseNameOf, pathWithTrailingSlashRoot) {
+        auto p1 = baseNameOf("/");
+        ASSERT_STREQ(std::string(p1).c_str(), "");
+    }
+
+    // XXX: according to the doc of `baseNameOf`, baseNameOf("/dir/") should return
```

Comment:
The behavior is consistent with what `basename` does in the shell so i guess we can just change the description of the function?